### ISSUE TITLE
bug: Directories scan never ends on v2.0.2

### ISSUE THREAD

**Issue by DrorDvash** [yogeshojha/rengine#1113]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I've updated reNgine to version 2.0.2 after I saw a closed issue where ffuf results were fixed and are now shown in the dashboard UI. This issue seems to be resolved, and the results are presented. However, the scan running never ends or keeps running over and over again.

Scan type: Subdomain Discovery, Port Scan, Directory and Files Search
![image](https://github.com/yogeshojha/rengine/assets/8413651/55cee238-ecce-4c96-b4ce-cd4c74da8830)

Current scan running for 23 hours (in v1.3.6 **the same** scan finished in 30-50 minutes)
![image](https://github.com/yogeshojha/rengine/assets/8413651/44855703-49ce-493d-81e0-1cc19d43dc76)

When looking at the results, I noticed that each subdomain was scanned multiple times for directory fuzzing (ffuf) instead of just once.
![image](https://github.com/yogeshojha/rengine/assets/8413651/1dfc0f7b-984b-4036-a991-64bc07a48961)

### Expected Behavior

Each domain should be scanned 1 time only with ffuf, and the scan should end correctly.

### Steps To Reproduce

1. Create Custom (quick) scan engine:
```yaml
subdomain_discovery: {
  'uses_tools': ['subfinder', 'ctfr', 'sublist3r', 'tlsx', 'oneforall', 'netlas'],
  'enable_http_crawl': true,
  'threads': 30,
  'timeout': 5,
}
http_crawl: {}
port_scan: {
  'enable_http_crawl': true,
  'timeout': 5,
  # 'exclude_ports': [],
  # 'exclude_subdomains': [],
  'ports': ['top-100'],
  'rate_limit': 150,
  'threads': 30,
  'passive': false,
  # 'use_naabu_config': false,
  # 'enable_nmap': true,
  # 'nmap_cmd': '',
  # 'nmap_script': '',
  # 'nmap_script_args': ''
}
dir_file_fuzz: {
  'auto_calibration': true,
  'enable_http_crawl': true,
  'rate_limit': 150,
  'extensions': ['html', 'php','git','yaml','conf','cnf','config','gz','env','log','db','mysql','bak','asp','aspx','txt','conf','sql','json','yml','pdf'],
  'follow_redirect': false,
  'max_time': 0,
  'match_http_status': [200, 204],
  'recursive_level': 2,
  'stop_on_error': false,
  'timeout': 5,
  'threads': 30,
  'wordlist_name': 'dicc'
}
screenshot: {
  'enable_http_crawl': true,
  'intensity': 'normal',
  'timeout': 10,
  'threads': 40
}

# custom_header: "Cookie: Test"
```

2. Start a new scan using the above engine.

### Environment

```markdown
- reNgine: v2.0.2
- OS: Ubuntu 22.04
- Python: Python 3.10.12
- Docker Engine: 24.0.7
- Docker Compose: v2.21.0
```

### Anything else?

If any logs are needed, please specify which ones and provide instructions on how to extract them for you. (I used `make logs`, but there are numerous lines.)

Thank you.

**Comment by AnonymousWP** [comment]
Did you update httpx via tools arsenal?

**Comment by DrorDvash** [comment]
Yes i did, already did update all tools links in the Dockerfile. (e.g. amass to v4)
![image](https://github.com/yogeshojha/rengine/assets/8413651/f474fb98-533c-471d-9f4e-1bddf08af135)

![image](https://github.com/yogeshojha/rengine/assets/8413651/b8ef9441-7093-4116-aeb6-ee5502b4b0e0)

so everything is up-to-date.

**Comment by AnonymousWP** [comment]
Thanks, I'm trying to reproduce the issue. If you want to post some detailed logs, check out https://github.com/yogeshojha/rengine/pull/994.

Are you running Docker Desktop on Windows perhaps? In that case you can also check logs directly in the container:

![image](https://github.com/yogeshojha/rengine/assets/50231698/3eb8de21-f010-4d1b-a5e7-c169c885e792)

**Comment by DrorDvash** [comment]
Well, i did `export DEBUG=1` and then `docker-compose restart web`, know i'm getting `502 Bad Gateway nginx/1.25.3`..i cannot see the dashboard anymore.

In addition, i think there is something causing error in the `make logs` command, every time i'm running it:
`error from daemon in stream: Error grabbing logs: invalid character 'l' after object key:value pair`
![image](https://github.com/yogeshojha/rengine/assets/8413651/6678b86b-0603-4ff1-8a7c-df0cf08a2139)
but that's another issue not related.

**Comment by psyray** [comment]
I also noticed this problem, **ffuf** is spawned over an over, but only one time.
If I kill the parent celery process and relaunch task it runs only once
Really strange problem.

**Comment by psyray** [comment]
> Well, i did `export DEBUG=1` and then `docker-compose restart web`, know i'm getting `502 Bad Gateway nginx/1.25.3`..i cannot see the dashboard anymore.
> 
> In addition, i think there is something causing error in the `make logs` command, every time i'm running it: `error from daemon in stream: Error grabbing logs: invalid character 'l' after object key:value pair` but that's another issue not related.

Weird, I'll do this a lot of time and no problem

**Comment by DrorDvash** [comment]
i have removed the `export DEBUG=1` and then make down && make up, now i can see the dashboard.
i'll try again

**Comment by psyray** [comment]
OK I think I've understood the problem.
FFUF launch is inside a loop of retrieved URLs, 
https://github.com/yogeshojha/rengine/blob/fd5a5e5faa1e289cbf421e8aaf6014452efaef1f/web/reNgine/tasks.py#L1631
So I think there's a problem somewhere in the URL retrieval
https://github.com/yogeshojha/rengine/blob/fd5a5e5faa1e289cbf421e8aaf6014452efaef1f/web/reNgine/tasks.py#L1620

I will try to debug.
Anyway source of the problem is the Default URL flag, see #1056 
We need to rework this part and some more checks to prevent non existing default URL

**Comment by psyray** [comment]
Found it, problem come from here
https://github.com/yogeshojha/rengine/blob/fd5a5e5faa1e289cbf421e8aaf6014452efaef1f/web/reNgine/tasks.py#L1683-L1684

Newly created endpoint are appended to the **urls** var.
As **urls** var is the loop var, at each newly created endpoint, script add another entry in the loop.
Recursive launch of ffuf

Don't know why this is here...
@AnonymousWP @yogeshojha 
Any idea ?
I think I could delete it

**Comment by AnonymousWP** [comment]
@psyray Nicely spotted, I was also thinking that there should be some infinite loop somewhere in the code due to a for-loop. Has this code always been present (I didn't bother checking)? Maybe with ocervell's PR. Anyway, I think you could delete and test locally, then see whether any errors arise and whether it fixes the issue or not.

**Comment by psyray** [comment]
> @psyray Nicely spotted, I was also thinking that there should be some infinite loop somewhere in the code due to a for-loop. Has this code always been present (I didn't bother checking)? Maybe with ocervell's PR. Anyway, I think you could delete and test locally, then see whether any errors arise and whether it fixes the issue or not.

It fixes, for sure.
I have also added other bug fix while debugging this one.
Currently testing

**Comment by DrorDvash** [comment]
I'm glad to see that you fixed the issue, and I would like to get the newest code releases + the issue fix, but I'm a little bit confused which branch should i stick with for now?
I have checked the **master** -> `web/reNgine/tasks.py` and i can see the issue you pointed to (`urls.append(endpoint.http_url)`) it is still here, not integrated in the master.
![image](https://github.com/yogeshojha/rengine/assets/8413651/c3ee5a3f-9a8c-490d-a7a9-19b91cb689e1)

i have checked the **2.1.0** -> `web/reNgine/tasks.py` - the same.

So, which branch has the latest commits + ffuf fix?

**Comment by psyray** [comment]
https://github.com/yogeshojha/rengine/pull/1120

**Comment by DrorDvash** [comment]
> #1120
> 
> Mine https://github.com/yogeshojha/rengine/tree/fix-recursive-ffuf-launch
> 
> Do a git pull and a checkout
> 
> ```
> git pull
> git checkout fix-recursive-ffuf-launch
> ```

yes I saw that branch but I've also seen more new commits from the very last days in the master / 2.1.0 branches, so I wanted to have the newest features / bug fixes in addition to the ffuf fix.

So there is no such branch currently?

**Comment by AnonymousWP** [comment]
You can switch to the 2.1.0 branch, as that contains all 2.1.0-related fixes.

**Comment by psyray** [comment]
> You can switch to the 2.1.0 branch, as that contains all 2.1.0-related fixes.

But not this one 😁

**Comment by DrorDvash** [comment]
> > You can switch to the 2.1.0 branch, as that contains all 2.1.0-related fixes.
> 
> But not this one 😁

you haven't merged ffuf fix (fix-recursive-ffuf-launch) to any other branch with the latest commits?

**Comment by psyray** [comment]
> > > You can switch to the 2.1.0 branch, as that contains all 2.1.0-related fixes.
> > 
> > 
> > But not this one 😁
> 
> you haven't merged ffuf fix (fix-recursive-ffuf-launch) to any other branch with the latest commits?

Nope, fix target master directly.
we can't wait release 2.1.0 to merge this one.

**Comment by scottnzuk** [comment]
I am seeing a very similar error to this on my Mac  M1

django.server | "GET /api/notifications/unread_count/?project_slug=default HTTP/1.0" 200 11
error from daemon in stream: Error grabbing logs: invalid character '\x00' looking for beginning of value

error from daemon in stream: Error grabbing logs: invalid character '\x00' looking for beginning of value
